{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b7b3c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import * \n",
    "from notes import NOTES \n",
    "from src.files.blast import BLASTFileGgKbase\n",
    "import glob\n",
    "import parasail \n",
    "import math\n",
    "import pyfaidx \n",
    "\n",
    "import networkx as nx\n",
    "from tqdm import tqdm \n",
    "%load_ext autoreload \n",
    "%autoreload 2\n",
    "\n",
    "# /groups/banfield/sequences/2014/16FT_4_MIN/raw.d/16ft_4_CZBZ.6237.3.40316_trim_clean.PE.1.fastq.gz\n",
    "\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_155357 is too low converage and has too many SNPs, not doable. \n",
    "# Nantong_Groundwater_SRR22387873_scaffold_346753\n",
    "# bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 in1=/groups/banfield/scratch/projects/environmental/RES/int/LDS/sandpiper_not_in_Logan/SRR22387873_trim_clean.PE.1.fastq.gz in2=/groups/banfield/scratch/projects/environmental/RES/int/LDS/sandpiper_not_in_Logan/SRR22387873_trim_clean.PE.2.fastq.gz ref=~/scaffold_346753.ref.fasta nodisk | shrinksam | sambam > ~/scaffold_346753.bam\n",
    "\n",
    "\n",
    "# makeblastdb -in contigs.fa -dbtype nucl -title contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f85a6b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_refs(df:pd.DataFrame, seed_contig_id:str='', data_dir='../data/bins/nantong_groundwater/'):\n",
    "    df.index = [re.search(r'scaffold_.+', contig_id, flags=re.DOTALL).group(0) for contig_id in df.index]\n",
    "    path = os.path.join(data_dir, f'{seed_contig_id}.ref.fasta')\n",
    "    FASTAFile.from_df(df).write(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "639ccd96",
   "metadata": {},
   "outputs": [],
   "source": [
    "CONTIGS_PATH = '../data/databases/nantong_groundwater/contigs.fa'\n",
    "\n",
    "def fai_get_sequences(contig_ids, path:str=CONTIGS_PATH, strands:list=None):\n",
    "    '''Read the sequences from the FASTA file.'''\n",
    "    f = pyfaidx.Fasta(path)\n",
    "    strands = np.array(['+'] * len(contig_ids)) if (strands is None) else strands\n",
    "    seqs = [str(f[contig_id]) for contig_id in contig_ids]\n",
    "    seqs = [reverse_complement(seq) if (strand == '-') else seq for (strand, seq) in zip(strands, seqs)]\n",
    "    return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b941f013",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_genome(input_dir:str, output_dir:str='../data/'):\n",
    "    contig_id_map = dict()\n",
    "    genome_id = os.path.basename(input_dir)\n",
    "    content = ''\n",
    "    for i, path in enumerate(sorted(glob.glob(os.path.join(input_dir, '*')))):\n",
    "        with open(path, 'r') as f:\n",
    "            lines = f.readlines()\n",
    "            contig_id_map[f'{genome_id}.{i + 1}'] = lines[0].split()[0].replace('>', '')\n",
    "            lines = [f'>{genome_id}.{i + 1}\\n'] + lines[1:]\n",
    "            content += ''.join(lines)\n",
    "\n",
    "    output_path = os.path.join(output_dir, f'{genome_id}.genome.fasta')\n",
    "    with open(output_path, 'w') as f:\n",
    "        f.write(content)\n",
    "    return output_path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "30ecb211",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_database(path:str='../data/databases/nantong_groundwater/taxonomy.tsv'):\n",
    "    # Contig name\tSize (bp)\tCoverage\tGC %\tTaxonomy winner\tWinner %\tSpecies winner\tSpecies winner %\tGenus winner\tGenus winner %\tOrder winner\tOrder winner %\tClass winner\tClass winner %\tPhylum winner\tPhylum winner %\tDomain winner\tDomain winner %\n",
    "    levels = ['species', 'genus', 'order', 'class', 'phylum', 'domain']\n",
    "    cols = ['contig_id', 'size', 'coverage', 'gc_percent', 'taxonomy_winner', 'taxonomy_winner_percent']\n",
    "    cols = cols + [col for level in levels for col in [f'{level}_winner', f'{level}_winner_percent']]\n",
    "    database_df = pd.read_csv(path, sep='\\t', names=cols, skiprows=1)\n",
    "    database_df = database_df.fillna('unknown')\n",
    "    return database_df.set_index('contig_id')\n",
    "\n",
    "database_df = load_database()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fd9d0335",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nantong_Groundwater_SRR22387873_scaffold_346753\n",
      "GC percent: 22.06%\n",
      "Length: 1047 bp\n"
     ]
    }
   ],
   "source": [
    "seed_contigs_df = pd.read_csv('../data/seed_contigs.csv', index_col=0)\n",
    "\n",
    "seed_contig_id = 'Nantong_Groundwater_SRR22387873_scaffold_346753'\n",
    "query_row = seed_contigs_df.loc[seed_contig_id]\n",
    "print(seed_contig_id)\n",
    "print(f'GC percent: {100 * query_row.gc_content:.2f}%')\n",
    "print(f'Length: {query_row.length} bp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b97a7797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "search_database: Found 9 with similar GC percent and coverage to Nantong_Groundwater_SRR22387873_scaffold_346753.\n",
      "search_database: 13.15 kb in candidate bin.\n"
     ]
    }
   ],
   "source": [
    "def search_database(contig_id, database_df=database_df, gc_percent_delta:float=3, coverage_delta:float=3, notes=NOTES, exclude_flagged:bool=True):\n",
    "    '''Search the database_df for contigs with similar coverage and GC content to the seed contig. '''\n",
    "    query_row = database_df.loc[contig_id]\n",
    "    mask = (database_df.gc_percent < query_row.gc_percent + gc_percent_delta) & (database_df.gc_percent > query_row.gc_percent - gc_percent_delta)\n",
    "    mask = mask & ((database_df.coverage < query_row.coverage + coverage_delta) & (database_df.coverage > query_row.coverage - coverage_delta))\n",
    "    \n",
    "    results_df = database_df[mask].copy()\n",
    "    results_df['note'] = results_df.index.map(notes)\n",
    "    results_df['note'] = results_df.note.fillna('none')\n",
    "    if exclude_flagged:\n",
    "        results_df = results_df[~results_df.note.str.contains('X')].copy()\n",
    "    results_df = results_df[results_df.domain_winner != 'Bacteria'].copy() # Exclude hits with Bacterial taxonomy winner.\n",
    "    results_df = results_df.sort_values('size', ascending=False)\n",
    "    results_df = results_df.loc[[contig_id] + [id_ for id_ in results_df.index if (id_ != contig_id)]].copy()\n",
    "    print(f'search_database: Found {len(results_df)} with similar GC percent and coverage to {contig_id}.')\n",
    "    print(f\"search_database: {results_df['size'].sum() / 1000:.2f} kb in candidate bin.\")\n",
    "    return results_df.copy()\n",
    "\n",
    "\n",
    "results_df = search_database(seed_contig_id)\n",
    "results_df['seq'] = fai_get_sequences(results_df.index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d3d5580b",
   "metadata": {},
   "outputs": [],
   "source": [
    "CONTIG_END_LENGTH = 200\n",
    "DATABASE_PATH = '/home/prichter/Documents/banfield/betazoid/data/databases/nantong_groundwater/contigs.fa'\n",
    "OUTPUT_DIR = '../data/bins/nantong_groundwater/'\n",
    "MIN_PERCENT_IDENTITY = 80\n",
    "MIN_ALIGNMENT_LENGTH = 80\n",
    "\n",
    "\n",
    "def _filter_hits(blast_df:pd.DataFrame, delta:int=3):\n",
    "    mask = (blast_df.query_border != blast_df.subject_border) # Get rid of overlaps between left borders and right borders. \n",
    "    mask = mask & (((blast_df.query_border == 'L') & (blast_df.query_lbd < delta)) | ((blast_df.query_border == 'R') & (blast_df.query_rbd < delta)))\n",
    "    mask = mask & (((blast_df.subject_border == 'L') & (blast_df.subject_lbd < delta)) | ((blast_df.subject_border == 'R') & (blast_df.subject_rbd < delta)))\n",
    "    print(f'_filter_hits: Retaining {mask.sum()} out of {len(blast_df)} hits.')\n",
    "    blast_df = blast_df[mask].copy()\n",
    "    return blast_df\n",
    "\n",
    "# Need some functions to account for merged contigs. \n",
    "get_scaffold_numbers = lambda contig_id : contig_id.split('_scaffold_')[-1].split('_')\n",
    "# has_shared_scaffold = lambda contig_id_1, contig_id_2 : len(np.intersect1d(get_scaffold_numbers(contig_id_1), get_scaffold_numbers(contig_id_2)))\n",
    "has_shared_scaffold = lambda row : len(np.intersect1d(get_scaffold_numbers(row.query_id), get_scaffold_numbers(row.subject_id))) > 0\n",
    "\n",
    "def blast_contig_ends(contigs_df:pd.DataFrame, name:str=None, max_n_mismatches:int=2):\n",
    "\n",
    "    input_path = os.path.join(OUTPUT_DIR, f'{name}.blast.in.fasta')\n",
    "    output_path = os.path.join(OUTPUT_DIR, f'{name}.blast.out.tsv')\n",
    "    input_df = [{'seq':row.seq[:CONTIG_END_LENGTH], 'id':f'{row.Index}.L'} for row in contigs_df.itertuples()]\n",
    "    input_df += [{'seq':row.seq[-CONTIG_END_LENGTH:], 'id':f'{row.Index}.R'} for row in contigs_df.itertuples()]\n",
    "    input_df = pd.DataFrame(input_df).set_index('id')\n",
    "    FASTAFile.from_df(input_df).write(input_path)\n",
    "\n",
    "    fields = 'qseqid sseqid qlen slen length mismatch gapopen qstart qend sstart send' #.split()\n",
    "    cmd = f'blastn -db {DATABASE_PATH} -query {input_path} -out {output_path} -perc_identity {MIN_PERCENT_IDENTITY} -outfmt \"6 {fields}\"'\n",
    "    \n",
    "    subprocess.run(cmd, shell=True, check=True)\n",
    "    blast_df = pd.read_csv(output_path, sep='\\t', names=fields.split())\n",
    "    blast_df = blast_df[blast_df.mismatch <= max_n_mismatches].copy()\n",
    "    # blast_df = blast_df[blast_df.qseqid.str.partition(\".\")[0] != blast_df.sseqid].copy() # Remove self-alignments.\n",
    "    blast_df = blast_df[blast_df.length > MIN_ALIGNMENT_LENGTH].copy() \n",
    "\n",
    "    # First load the subject sequences and clean up. \n",
    "    # Convert from one-indexed inclusive bounds to zero-indexed, upper-exclusive bounds. \n",
    "\n",
    "    blast_df['subject_id'] = blast_df.sseqid\n",
    "    blast_df['subject_strand'] = np.where(blast_df.send < blast_df.sstart, '-', '+')\n",
    "    blast_df['subject_length'] = blast_df.slen\n",
    "    blast_df['subject_seq'] = fai_get_sequences(blast_df.subject_id, strands=blast_df.subject_strand.values)\n",
    "    blast_df['subject_start'] = np.where(blast_df.subject_strand == '+', blast_df.sstart - 1, blast_df.slen - blast_df.sstart)\n",
    "    blast_df['subject_end'] = np.where(blast_df.subject_strand == '+', blast_df.send, blast_df.slen - blast_df.send + 1)\n",
    "    blast_df['subject_rbd'] = blast_df.subject_length - blast_df.subject_end\n",
    "    blast_df['subject_lbd'] = blast_df.subject_start\n",
    "    blast_df['subject_border'] = np.where(blast_df.subject_lbd < blast_df.subject_rbd, 'L', 'R')\n",
    "\n",
    "    blast_df['query_id'] = blast_df.qseqid.str.partition(\".\")[0]\n",
    "    blast_df['query_seq'] = blast_df.query_id.map(contigs_df.seq)\n",
    "    blast_df['query_length'] = blast_df.query_seq.apply(len)\n",
    "    blast_df['query_border'] = [re.search(r'\\.(R|L)', id_).group(1) for id_ in blast_df.qseqid]\n",
    "    blast_df['query_start'] = np.where(blast_df.query_border == 'R', (blast_df.query_length - CONTIG_END_LENGTH) + (blast_df.qstart - 1), blast_df.qstart - 1) # If query match is on the right border, the start of the alignment is (150 - qstart) from the right end. \n",
    "    blast_df['query_end'] = np.where((blast_df.query_border == 'R'),  (blast_df.query_length - CONTIG_END_LENGTH) + blast_df.qend, blast_df.qend) # If query match is on the right border, use the end of the alignment is (150 - qend) from the right end. \n",
    "    blast_df['query_rbd'] = blast_df.query_length - blast_df.query_end\n",
    "    blast_df['query_lbd'] = blast_df.query_start\n",
    "\n",
    "    blast_df = blast_df[~blast_df.apply(has_shared_scaffold, axis=1)].copy() # Remove self-hits.\n",
    "\n",
    "    blast_df = blast_df.rename(columns={'gapopen':'n_gaps', 'mismatch':'n_mismatches', 'length':'alignment_length'})\n",
    "    cols = ['query_id', 'subject_id', 'query_length', 'subject_length', 'subject_strand'] \n",
    "    cols += ['query_border', 'query_lbd', 'query_rbd', 'subject_border', 'subject_lbd', 'subject_rbd']\n",
    "    cols += ['query_start', 'query_end', 'subject_start', 'subject_end']\n",
    "    cols += ['n_gaps', 'n_mismatches', 'alignment_length', 'query_seq', 'subject_seq']\n",
    "    return blast_df[cols].copy().reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ffd3e729",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_contigs(blast_df:pd.DataFrame, database_df:pd.DataFrame=database_df):\n",
    "    get_scaffold_number = lambda contig_id : contig_id.split('_scaffold_')[-1] #.split('_')\n",
    "    get_scaffold_prefix = lambda contig_id : contig_id.split('_scaffold_')[0]\n",
    "\n",
    "    df = list()\n",
    "    for row in blast_df.itertuples():\n",
    "        row_ = {'parent_contig_ids':[row.query_id, row.subject_id], 'parent_lengths':[row.query_length, row.subject_length], 'alignment_length':row.alignment_length}\n",
    "        row_['parent_coverages'] = database_df.loc[[row.query_id, row.subject_id]].coverage.tolist()\n",
    "        row_['parent_gc_percents'] = database_df.loc[[row.query_id, row.subject_id]].gc_percent.tolist()\n",
    "\n",
    "        if row.query_border == 'R':\n",
    "            row_['seq'] = row.query_seq + row.subject_seq[row.subject_end:]\n",
    "            row_['contig_id'] = get_scaffold_prefix(row.query_id) + '_scaffold_' + get_scaffold_number(row.query_id) + '_' + get_scaffold_number(row.subject_id)\n",
    "        elif row.query_border == 'L':\n",
    "            row_['seq'] = row.subject_seq[:row.subject_start] + row.query_seq\n",
    "            row_['contig_id'] = get_scaffold_prefix(row.query_id) + '_scaffold_' + get_scaffold_number(row.subject_id) + '_' + get_scaffold_number(row.query_id)\n",
    "        \n",
    "        L = sum(row_['parent_lengths'])\n",
    "        row_['coverage'] = np.mean([(l /L)  * gc for l, gc in zip(row_['parent_lengths'], row_['parent_gc_percents'])]) # This will not be exact because of the overlap.\n",
    "        row_['gc_percent'] = 100 * (row_['seq'].count('G') + row_['seq'].count('C')) / len(row_['seq'])\n",
    "        df.append(row_)\n",
    "    df = pd.DataFrame(df)\n",
    "    df['length'] = df.seq.apply(len)\n",
    "    \n",
    "    # print(df.query_length + df.subject_length - df.alignment_length)\n",
    "    return df.set_index('contig_id', drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "52422697",
   "metadata": {},
   "outputs": [],
   "source": [
    "# blast_df = _filter_hits(blast_contig_ends(results_df, name='scaffold_346753'))\n",
    "# blast_df = blast_contig_ends(results_df, name='scaffold_346753')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "36cb8418",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -db /home/prichter/Documents/banfield/betazoid/data/databases/nantong_groundwater/contigs.fa -query ../data/bins/nantong_groundwater/scaffold_2044.2.gene.fasta -out ../data/bins/nantong_groundwater/scaffold_2044.2.blast.tsv -outfmt \"6 qseqid sseqid qlen slen length mismatch gapopen qstart qend sstart send pident bitscore evalue\" -perc_identity 10 -evalue 10 -gapopen 5  -gapextend 2 -word_size 7 -penalty -1\n"
     ]
    }
   ],
   "source": [
    "fields = 'qseqid sseqid qlen slen length mismatch gapopen qstart qend sstart send pident bitscore evalue' #.split()\n",
    "input_path = '../data/bins/nantong_groundwater/scaffold_2044.2.gene.fasta'\n",
    "output_path = '../data/bins/nantong_groundwater/scaffold_2044.2.blast.tsv'\n",
    "cmd = f'blastn -db {DATABASE_PATH} -query {input_path} -out {output_path} -outfmt \"6 {fields}\" -perc_identity 10 -evalue 10 -gapopen 5  -gapextend 2 -word_size 7 -penalty -1'\n",
    "print(cmd)\n",
    "\n",
    "blast_df = pd.read_csv(output_path, sep='\\t', names=fields.split())\n",
    "blast_df = blast_df[blast_df.sseqid != 'Nantong_Groundwater_SRR22387873_scaffold_2044'].copy() # Remove self-hits.\n",
    "blast_df = blast_df[blast_df.sseqid != 'Nantong_Groundwater_SRR22387873_scaffold_1350'].copy() # Remove hits to the other known Betazoid.  \n",
    "blast_df = blast_df[blast_df.slen < 25000].copy() # Exclude contig hits that are too long. \n",
    "\n",
    "subject_contig_ids = blast_df.sseqid.unique()\n",
    "blast_df['subject_seq'] = blast_df.sseqid.map(dict(zip(subject_contig_ids, fai_get_sequences(subject_contig_ids))))\n",
    "\n",
    "get_gc_percent = lambda seq : 100 * (seq.count('G') + seq.count('C')) / len(seq)\n",
    "blast_df['subject_gc_percent'] = blast_df.subject_seq.apply(get_gc_percent)\n",
    "blast_df = blast_df[blast_df.subject_gc_percent < 25].copy()\n",
    "\n",
    "blast_df['taxonomy_winner'] = blast_df.sseqid.map(database_df.taxonomy_winner)\n",
    "blast_df['domain_winner'] = blast_df.sseqid.map(database_df.domain_winner)\n",
    "# blast_df['taxonomy_winner_percent'] = blast_df.sseqid.map(database_df.taxonomy_winner_percent)\n",
    "# blast_df['taxonomy_winner_percent'] = np.where(blast_df.taxonomy_winner_percent == 'unknown', 0, blast_df.taxonomy_winner_percent).astype(float)\n",
    "blast_df = blast_df[(blast_df.taxonomy_winner == 'unknown')].copy()\n",
    "\n",
    "get_scaffold_number = lambda contig_id : int(contig_id.split('_scaffold_')[-1])\n",
    "blast_df['subject_contig_number'] = blast_df.sseqid.apply(get_scaffold_number)\n",
    "blast_df = blast_df[blast_df.length > 50].copy()\n",
    "\n",
    "exclude_scaffold_numbers = [43870, 253652, 370601, 267053, 134067, 30515, 38970, 106191, 3756, 4040, 5622, 9140, 10230, 10553, 11878, 12630,13505, 13849, 13997, 21904, 2414, 82343, 9811, 24147, 24838, 28963, 47738]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "7c1d2df9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>qlen</th>\n",
       "      <th>slen</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>pident</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>evalue</th>\n",
       "      <th>subject_seq</th>\n",
       "      <th>subject_gc_percent</th>\n",
       "      <th>taxonomy_winner</th>\n",
       "      <th>domain_winner</th>\n",
       "      <th>subject_contig_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8573</th>\n",
       "      <td>scaffold_2044.2_20</td>\n",
       "      <td>Nantong_Groundwater_SRR22387873_scaffold_18414</td>\n",
       "      <td>1833</td>\n",
       "      <td>4471</td>\n",
       "      <td>65</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>1427</td>\n",
       "      <td>1491</td>\n",
       "      <td>4304</td>\n",
       "      <td>4240</td>\n",
       "      <td>69.231</td>\n",
       "      <td>41.2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>CTTGCTTTTACTAATTTTTCGTCAGTAAAAATGATACCATTTCCGT...</td>\n",
       "      <td>24.021472</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>18414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  qseqid                                          sseqid  \\\n",
       "8573  scaffold_2044.2_20  Nantong_Groundwater_SRR22387873_scaffold_18414   \n",
       "\n",
       "      qlen  slen  length  mismatch  gapopen  qstart  qend  sstart  send  \\\n",
       "8573  1833  4471      65        20        0    1427  1491    4304  4240   \n",
       "\n",
       "      pident  bitscore  evalue  \\\n",
       "8573  69.231      41.2    0.52   \n",
       "\n",
       "                                            subject_seq  subject_gc_percent  \\\n",
       "8573  CTTGCTTTTACTAATTTTTCGTCAGTAAAAATGATACCATTTCCGT...           24.021472   \n",
       "\n",
       "     taxonomy_winner domain_winner  subject_contig_number  \n",
       "8573         unknown       unknown                  18414  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_df[(blast_df.bitscore > 40) & (~blast_df.subject_contig_number.isin(exclude_scaffold_numbers))].sort_values('slen').drop_duplicates('sseqid')\n",
    "\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_130073 looks super promising. \n",
    "# Nantong_Groundwater_SRR22387873_scaffold_18414 might also be a good candidate.\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_119008\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_90561 maybe another DNA polymerase.\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_239429 could be another DNA polymerase\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_300679 is very promising. \n",
    "# Nantong_Groundwater_SRR22387873_scaffold_290644 is probably another DNA polymerase.\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_183905 another DNA polymerase?\n",
    "# Nantong_Groundwater_SRR22387873_scaffold_241552\n",
    "blast_df[blast_df.sseqid == 'Nantong_Groundwater_SRR22387873_scaffold_18414']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bdbe8651",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_refs(results_df, seed_contig_id='scaffold_346753')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15dcc182",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace_contig_ids = dict()\n",
    "# replace_contig_ids['Nantong_Groundwater_SRR22387873_scaffold_377775'] = 'Nantong_Groundwater_SRR22387873_scaffold_103075_377775'\n",
    "# replace_contig_ids['Nantong_Groundwater_SRR22387873_scaffold_358410'] = 'Nantong_Groundwater_SRR22387873_scaffold_153884_358410'\n",
    "# replace_contig_ids['Nantong_Groundwater_SRR22387873_scaffold_335499'] = 'Nantong_Groundwater_SRR22387873_scaffold_45379_335499_207268'\n",
    "\n",
    "# for contig_id, new_contig_id in replace_contig_ids.items():\n",
    "#     results_df.loc[contig_id, 'size'] = new_contigs_df.loc[new_contig_id, 'length']\n",
    "#     results_df.loc[contig_id, 'gc_percent'] = new_contigs_df.loc[new_contig_id, 'gc_percent']\n",
    "#     results_df.loc[contig_id, 'coverage'] = new_contigs_df.loc[new_contig_id, 'coverage']\n",
    "#     results_df.loc[contig_id, 'seq'] = new_contigs_df.loc[new_contig_id, 'seq']\n",
    "#     results_df.rename(index={contig_id:new_contig_id}, inplace=True)\n",
    "\n",
    "# FASTAFile.from_df(results_df).write(f\"../data/bins/nantong_groundwater/{query_row['id']}.ref.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c19bde99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(13147)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df.sort_values('size')\n",
    "results_df['size'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8eeb433d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2797\n",
      "3860\n",
      "1865\n"
     ]
    }
   ],
   "source": [
    "print(database_df.loc['Nantong_Groundwater_SRR22387873_scaffold_45379']['size'])\n",
    "print(database_df.loc['Nantong_Groundwater_SRR22387873_scaffold_335499']['size'] + database_df.loc['Nantong_Groundwater_SRR22387873_scaffold_45379']['size'])\n",
    "\n",
    "print(database_df.loc['Nantong_Groundwater_SRR22387873_scaffold_103075']['size'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4133bb2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_contig_ends(contigs_df, min_percent_identity:float=0.98, min_alignment_length:int=15, length:int=150, gap_open:int=20, gap_extend:int=1):\n",
    "    contigs_df['reverse_seq'] = [reverse_complement(seq) for seq in contigs_df.seq] \n",
    "    contigs_df['seq_l'] = [seq[:length] for seq in contigs_df.seq]\n",
    "    contigs_df['seq_r'] = [seq[-length:] for seq in contigs_df.seq]\n",
    "    contigs_df['reverse_seq_l'] = [seq[:length] for seq in contigs_df.reverse_seq]\n",
    "    contigs_df['reverse_seq_r'] = [seq[-length:] for seq in contigs_df.reverse_seq]\n",
    "\n",
    "    matrix = parasail.matrix_create('ACGT', 2, -3) # + 2 match score, -3 mismatch score. \n",
    "    graph = nx.DiGraph()\n",
    "\n",
    "    rows = list(contigs_df.itertuples())\n",
    "    for a, b in tqdm(itertools.product(rows, rows), desc='align_contig_ends'):\n",
    "        if a.Index == b.Index:\n",
    "            continue \n",
    "\n",
    "        for loc_a, loc_b in [('seq_r', 'seq_l'), ('seq_r', 'reverse_seq_l'), ('reverse_seq_r', 'seq_l')]:\n",
    "            alignment = parasail.sw_trace_striped_16(getattr(a, loc_a), getattr(b, loc_b), gap_open, gap_extend, matrix) # Smith-Waterman local alignment. \n",
    "            n_matches = sum(x == y for x, y in zip(alignment.traceback.query, alignment.traceback.ref) if ((x != '-') and (y != '-')))\n",
    "            n = sum(1 for x, y in zip(alignment.traceback.query, alignment.traceback.ref) if ((x != '-') and (y != '-')))\n",
    "            # print(alignment.traceback.comp)\n",
    "            if ((n_matches / n) > min_percent_identity) and (n > min_alignment_length):\n",
    "                graph.add_edge(a.Index, b.Index, percent_identity=(n_matches / n), alignment_length=n, alignment=f'{alignment.traceback.query}\\n{alignment.traceback.ref}')\n",
    "                \n",
    "    return graph "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "83985e8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "align_contig_ends: 0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "align_contig_ends: 81it [00:00, 7327.64it/s]\n"
     ]
    }
   ],
   "source": [
    "graph = align_contig_ends(results_df)\n",
    "# fig, ax = plt.subplots(figsize=(6, 6))\n",
    "# pos = nx.spring_layout(graph)\n",
    "# nx.draw_networkx_nodes(graph, pos, ax=ax, node_size=100)\n",
    "# nx.draw_networkx_edges(graph, pos, ax=ax, width=0.5)\n",
    "\n",
    "# for _, spine in ax.spines.items():\n",
    "#     spine.set_visible(False)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2caa3d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "for u, v, data in graph.edges(data=True):\n",
    "    print(u, v)\n",
    "    print(data['alignment'])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "399ad276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df['size'].sum()\n",
    "len(results_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d789c981",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>size</th>\n",
       "      <th>coverage</th>\n",
       "      <th>gc_percent</th>\n",
       "      <th>taxonomy_winner</th>\n",
       "      <th>taxonomy_winner_percent</th>\n",
       "      <th>species_winner</th>\n",
       "      <th>species_winner_percent</th>\n",
       "      <th>genus_winner</th>\n",
       "      <th>genus_winner_percent</th>\n",
       "      <th>order_winner</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum_winner_percent</th>\n",
       "      <th>domain_winner</th>\n",
       "      <th>domain_winner_percent</th>\n",
       "      <th>note</th>\n",
       "      <th>seq</th>\n",
       "      <th>reverse_seq</th>\n",
       "      <th>seq_l</th>\n",
       "      <th>seq_r</th>\n",
       "      <th>reverse_seq_l</th>\n",
       "      <th>reverse_seq_r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_346753</th>\n",
       "      <td>1047</td>\n",
       "      <td>23.07</td>\n",
       "      <td>22.06</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Seed contig with DNA polymerase B.</td>\n",
       "      <td>TTTATACGAATCTATAAAATATTTTGAGGAAATAATTAATAGTTTA...</td>\n",
       "      <td>ATTTCAAATTTATCAGTTTTATGATTATATCCTTCGATATAATAAG...</td>\n",
       "      <td>TTTATACGAATCTATAAAATATTTTGAGGAAATAATTAATAGTTTA...</td>\n",
       "      <td>TACTGACGCAATAATAACATCCGAAAAATTACCGACAGGAACAGGT...</td>\n",
       "      <td>ATTTCAAATTTATCAGTTTTATGATTATATCCTTCGATATAATAAG...</td>\n",
       "      <td>CCTGTTATTGTTTCACAATTAATTGAAGGAATAAATTTATTTAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_338320</th>\n",
       "      <td>1059</td>\n",
       "      <td>24.93</td>\n",
       "      <td>21.72</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No annotations.</td>\n",
       "      <td>ATGCACCAACACCTGAAGAAGCATTAAAAAATACATTCACTTTAGT...</td>\n",
       "      <td>CTATTTGTTTTTGAATTAATCCACTTATACTTTCAATTTCTAAACC...</td>\n",
       "      <td>ATGCACCAACACCTGAAGAAGCATTAAAAAATACATTCACTTTAGT...</td>\n",
       "      <td>ATTTTAGTAGTAATTCTTTAGATAAATCAATTTCTTTTACAGATTA...</td>\n",
       "      <td>CTATTTGTTTTTGAATTAATCCACTTATACTTTCAATTTCTAAACC...</td>\n",
       "      <td>CTGAAAGGAGAGGATTAAATTTTGAAATTTTTGCGGTTATAACACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_300679</th>\n",
       "      <td>1119</td>\n",
       "      <td>22.92</td>\n",
       "      <td>23.68</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No annotations.</td>\n",
       "      <td>GTAAAAAACATAAAAAATATACAGGTTCACGATATTAAATGAACCC...</td>\n",
       "      <td>CTCATTGCGTAAAGATAATAATATCCGTGTTTAAAAAAAAGATTGT...</td>\n",
       "      <td>GTAAAAAACATAAAAAATATACAGGTTCACGATATTAAATGAACCC...</td>\n",
       "      <td>ATCCTTTTCAATCACAATGCGAATATTCAATAGAAAAAGGATATAA...</td>\n",
       "      <td>CTCATTGCGTAAAGATAATAATATCCGTGTTTAAAAAAAAGATTGT...</td>\n",
       "      <td>AAATTAAAAAAGCTAAAGCTGTCATTGATAAGGTTTTACCTTTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_290644</th>\n",
       "      <td>1137</td>\n",
       "      <td>22.56</td>\n",
       "      <td>22.69</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Has another type B DNA polymerase. Also a Beta...</td>\n",
       "      <td>AAATAGACAGATTAAATTTAGAGATTCATACGCATTAATGCCGAAA...</td>\n",
       "      <td>TGTTTTAATTTTTCGGATGTAATAATAGCATCAGTATCACAATAAG...</td>\n",
       "      <td>AAATAGACAGATTAAATTTAGAGATTCATACGCATTAATGCCGAAA...</td>\n",
       "      <td>AAATCTTGATTCTAATCATATTATTCCATCGTTATCATACAGGGTT...</td>\n",
       "      <td>TGTTTTAATTTTTCGGATGTAATAATAGCATCAGTATCACAATAAG...</td>\n",
       "      <td>TGTCTTAAACACGTAAATATTCTTTTATCAGGATTATCAAAATCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_267693</th>\n",
       "      <td>1182</td>\n",
       "      <td>22.08</td>\n",
       "      <td>24.20</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.75</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.75</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>0.75</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.75</td>\n",
       "      <td>One hypothetical protein assigned Orange Borg ...</td>\n",
       "      <td>ATAAACCAATATCAGAACTAACTCAGCAAAATACTATTGGTATGCA...</td>\n",
       "      <td>TGTTTGATATGCTTTAATAACAATATTCTCTATTTGCTCTCTGTCA...</td>\n",
       "      <td>ATAAACCAATATCAGAACTAACTCAGCAAAATACTATTGGTATGCA...</td>\n",
       "      <td>GATAAAGAGACGAAAATAGATGAAAATCTTAAAGAGAAAATGGATA...</td>\n",
       "      <td>TGTTTGATATGCTTTAATAACAATATTCTCTATTTGCTCTCTGTCA...</td>\n",
       "      <td>AATTTTTTAATATATTTCATTTAATCATCCTTTTTATTTTTTTAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_241552</th>\n",
       "      <td>1241</td>\n",
       "      <td>21.27</td>\n",
       "      <td>23.05</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No annotations.</td>\n",
       "      <td>GCATTAATTAAATCAACAGCTAATCTTTTAATTTGAATATCGTAAT...</td>\n",
       "      <td>ATATATGTTCATATTTGGAATTTTTGGATGTATTTTAGGATATATT...</td>\n",
       "      <td>GCATTAATTAAATCAACAGCTAATCTTTTAATTTGAATATCGTAAT...</td>\n",
       "      <td>TATTGTATCTTCTCGATTCATTTTTAACACTGCCTAATTAATATGT...</td>\n",
       "      <td>ATATATGTTCATATTTGGAATTTTTGGATGTATTTTAGGATATATT...</td>\n",
       "      <td>AAAAAAAGATAATAAATATTTAGTTTTGAATATTGATGATGAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_231049</th>\n",
       "      <td>1267</td>\n",
       "      <td>22.97</td>\n",
       "      <td>21.31</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No annotations.</td>\n",
       "      <td>ATCATCATAACAACATATAACATCCATTTCATCTATCTCTTCACCT...</td>\n",
       "      <td>GAAAGAGAAAGTAGAAAAGAAATTTGATGGAAATGTAGTTGTATGG...</td>\n",
       "      <td>ATCATCATAACAACATATAACATCCATTTCATCTATCTCTTCACCT...</td>\n",
       "      <td>ATTTTTCATTTTATTTTCCTTCTTTTATTATATTCAACACAAACAA...</td>\n",
       "      <td>GAAAGAGAAAGTAGAAAAGAAATTTGATGGAAATGTAGTTGTATGG...</td>\n",
       "      <td>GATGGAGTGAAGAAGTATTTTGTGTTAGATGGTTGTTGGAGATGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_58476</th>\n",
       "      <td>2464</td>\n",
       "      <td>22.22</td>\n",
       "      <td>24.72</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.9</td>\n",
       "      <td>One hypothetical protein annotated as belongin...</td>\n",
       "      <td>AGGAAAAGACGGATTAAATGGTATTGACGGAAAAGATGTTGACCCT...</td>\n",
       "      <td>CTCCAATTTTAAAAAATCAATATCAGCTATTTTAATTTCTATTCCT...</td>\n",
       "      <td>AGGAAAAGACGGATTAAATGGTATTGACGGAAAAGATGTTGACCCT...</td>\n",
       "      <td>ATTAAAATAATAATGAATAAAGAAACAAAAGAAAAATTATTTACCG...</td>\n",
       "      <td>CTCCAATTTTAAAAAATCAATATCAGCTATTTTAATTTCTATTCCT...</td>\n",
       "      <td>TCTTGTTTTTTATTAACAGAAGATAAATTATCAATAGCTACAACAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_51309</th>\n",
       "      <td>2631</td>\n",
       "      <td>25.77</td>\n",
       "      <td>24.90</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.6</td>\n",
       "      <td>Has two hypothetical proteins annotated with B...</td>\n",
       "      <td>GACAAGATAATAAAAGACGTAACTTGCAAAAAAAGATAAAATAAAA...</td>\n",
       "      <td>AACTGTAAATGTTGTAGTGGCTGTTCCATCTCGTCTGTAATACTCA...</td>\n",
       "      <td>GACAAGATAATAAAAGACGTAACTTGCAAAAAAAGATAAAATAAAA...</td>\n",
       "      <td>GATAATACAGTTAATATTCAAAATAAACAAATTGAGATATTAAATA...</td>\n",
       "      <td>AACTGTAAATGTTGTAGTGGCTGTTCCATCTCGTCTGTAATACTCA...</td>\n",
       "      <td>GTAAGAGTAACGAAATTATTATACCAAAAACATCTATGTCTTAATT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 size  coverage  gc_percent taxonomy_winner  \\\n",
       "scaffold_346753  1047     23.07       22.06         unknown   \n",
       "scaffold_338320  1059     24.93       21.72         unknown   \n",
       "scaffold_300679  1119     22.92       23.68         unknown   \n",
       "scaffold_290644  1137     22.56       22.69         unknown   \n",
       "scaffold_267693  1182     22.08       24.20         unknown   \n",
       "scaffold_241552  1241     21.27       23.05         unknown   \n",
       "scaffold_231049  1267     22.97       21.31         unknown   \n",
       "scaffold_58476   2464     22.22       24.72         unknown   \n",
       "scaffold_51309   2631     25.77       24.90         unknown   \n",
       "\n",
       "                taxonomy_winner_percent species_winner species_winner_percent  \\\n",
       "scaffold_346753                 unknown        unknown                    1.0   \n",
       "scaffold_338320                 unknown        unknown                    1.0   \n",
       "scaffold_300679                 unknown        unknown                    1.0   \n",
       "scaffold_290644                 unknown        unknown                    1.0   \n",
       "scaffold_267693                 unknown        unknown                   0.75   \n",
       "scaffold_241552                 unknown        unknown                    1.0   \n",
       "scaffold_231049                 unknown        unknown                    1.0   \n",
       "scaffold_58476                  unknown        unknown                    0.9   \n",
       "scaffold_51309                  unknown        unknown                    0.6   \n",
       "\n",
       "                genus_winner genus_winner_percent order_winner  ...  \\\n",
       "scaffold_346753      unknown                  1.0      unknown  ...   \n",
       "scaffold_338320      unknown                  1.0      unknown  ...   \n",
       "scaffold_300679      unknown                  1.0      unknown  ...   \n",
       "scaffold_290644      unknown                  1.0      unknown  ...   \n",
       "scaffold_267693      unknown                 0.75      unknown  ...   \n",
       "scaffold_241552      unknown                  1.0      unknown  ...   \n",
       "scaffold_231049      unknown                  1.0      unknown  ...   \n",
       "scaffold_58476       unknown                  0.9      unknown  ...   \n",
       "scaffold_51309       unknown                  0.6      unknown  ...   \n",
       "\n",
       "                phylum_winner_percent domain_winner domain_winner_percent  \\\n",
       "scaffold_346753                   1.0       unknown                   1.0   \n",
       "scaffold_338320                   1.0       unknown                   1.0   \n",
       "scaffold_300679                   1.0       unknown                   1.0   \n",
       "scaffold_290644                   1.0       unknown                   1.0   \n",
       "scaffold_267693                  0.75       unknown                  0.75   \n",
       "scaffold_241552                   1.0       unknown                   1.0   \n",
       "scaffold_231049                   1.0       unknown                   1.0   \n",
       "scaffold_58476                    0.9       unknown                   0.9   \n",
       "scaffold_51309                    0.6       unknown                   0.6   \n",
       "\n",
       "                                                              note  \\\n",
       "scaffold_346753                 Seed contig with DNA polymerase B.   \n",
       "scaffold_338320                                    No annotations.   \n",
       "scaffold_300679                                    No annotations.   \n",
       "scaffold_290644  Has another type B DNA polymerase. Also a Beta...   \n",
       "scaffold_267693  One hypothetical protein assigned Orange Borg ...   \n",
       "scaffold_241552                                    No annotations.   \n",
       "scaffold_231049                                    No annotations.   \n",
       "scaffold_58476   One hypothetical protein annotated as belongin...   \n",
       "scaffold_51309   Has two hypothetical proteins annotated with B...   \n",
       "\n",
       "                                                               seq  \\\n",
       "scaffold_346753  TTTATACGAATCTATAAAATATTTTGAGGAAATAATTAATAGTTTA...   \n",
       "scaffold_338320  ATGCACCAACACCTGAAGAAGCATTAAAAAATACATTCACTTTAGT...   \n",
       "scaffold_300679  GTAAAAAACATAAAAAATATACAGGTTCACGATATTAAATGAACCC...   \n",
       "scaffold_290644  AAATAGACAGATTAAATTTAGAGATTCATACGCATTAATGCCGAAA...   \n",
       "scaffold_267693  ATAAACCAATATCAGAACTAACTCAGCAAAATACTATTGGTATGCA...   \n",
       "scaffold_241552  GCATTAATTAAATCAACAGCTAATCTTTTAATTTGAATATCGTAAT...   \n",
       "scaffold_231049  ATCATCATAACAACATATAACATCCATTTCATCTATCTCTTCACCT...   \n",
       "scaffold_58476   AGGAAAAGACGGATTAAATGGTATTGACGGAAAAGATGTTGACCCT...   \n",
       "scaffold_51309   GACAAGATAATAAAAGACGTAACTTGCAAAAAAAGATAAAATAAAA...   \n",
       "\n",
       "                                                       reverse_seq  \\\n",
       "scaffold_346753  ATTTCAAATTTATCAGTTTTATGATTATATCCTTCGATATAATAAG...   \n",
       "scaffold_338320  CTATTTGTTTTTGAATTAATCCACTTATACTTTCAATTTCTAAACC...   \n",
       "scaffold_300679  CTCATTGCGTAAAGATAATAATATCCGTGTTTAAAAAAAAGATTGT...   \n",
       "scaffold_290644  TGTTTTAATTTTTCGGATGTAATAATAGCATCAGTATCACAATAAG...   \n",
       "scaffold_267693  TGTTTGATATGCTTTAATAACAATATTCTCTATTTGCTCTCTGTCA...   \n",
       "scaffold_241552  ATATATGTTCATATTTGGAATTTTTGGATGTATTTTAGGATATATT...   \n",
       "scaffold_231049  GAAAGAGAAAGTAGAAAAGAAATTTGATGGAAATGTAGTTGTATGG...   \n",
       "scaffold_58476   CTCCAATTTTAAAAAATCAATATCAGCTATTTTAATTTCTATTCCT...   \n",
       "scaffold_51309   AACTGTAAATGTTGTAGTGGCTGTTCCATCTCGTCTGTAATACTCA...   \n",
       "\n",
       "                                                             seq_l  \\\n",
       "scaffold_346753  TTTATACGAATCTATAAAATATTTTGAGGAAATAATTAATAGTTTA...   \n",
       "scaffold_338320  ATGCACCAACACCTGAAGAAGCATTAAAAAATACATTCACTTTAGT...   \n",
       "scaffold_300679  GTAAAAAACATAAAAAATATACAGGTTCACGATATTAAATGAACCC...   \n",
       "scaffold_290644  AAATAGACAGATTAAATTTAGAGATTCATACGCATTAATGCCGAAA...   \n",
       "scaffold_267693  ATAAACCAATATCAGAACTAACTCAGCAAAATACTATTGGTATGCA...   \n",
       "scaffold_241552  GCATTAATTAAATCAACAGCTAATCTTTTAATTTGAATATCGTAAT...   \n",
       "scaffold_231049  ATCATCATAACAACATATAACATCCATTTCATCTATCTCTTCACCT...   \n",
       "scaffold_58476   AGGAAAAGACGGATTAAATGGTATTGACGGAAAAGATGTTGACCCT...   \n",
       "scaffold_51309   GACAAGATAATAAAAGACGTAACTTGCAAAAAAAGATAAAATAAAA...   \n",
       "\n",
       "                                                             seq_r  \\\n",
       "scaffold_346753  TACTGACGCAATAATAACATCCGAAAAATTACCGACAGGAACAGGT...   \n",
       "scaffold_338320  ATTTTAGTAGTAATTCTTTAGATAAATCAATTTCTTTTACAGATTA...   \n",
       "scaffold_300679  ATCCTTTTCAATCACAATGCGAATATTCAATAGAAAAAGGATATAA...   \n",
       "scaffold_290644  AAATCTTGATTCTAATCATATTATTCCATCGTTATCATACAGGGTT...   \n",
       "scaffold_267693  GATAAAGAGACGAAAATAGATGAAAATCTTAAAGAGAAAATGGATA...   \n",
       "scaffold_241552  TATTGTATCTTCTCGATTCATTTTTAACACTGCCTAATTAATATGT...   \n",
       "scaffold_231049  ATTTTTCATTTTATTTTCCTTCTTTTATTATATTCAACACAAACAA...   \n",
       "scaffold_58476   ATTAAAATAATAATGAATAAAGAAACAAAAGAAAAATTATTTACCG...   \n",
       "scaffold_51309   GATAATACAGTTAATATTCAAAATAAACAAATTGAGATATTAAATA...   \n",
       "\n",
       "                                                     reverse_seq_l  \\\n",
       "scaffold_346753  ATTTCAAATTTATCAGTTTTATGATTATATCCTTCGATATAATAAG...   \n",
       "scaffold_338320  CTATTTGTTTTTGAATTAATCCACTTATACTTTCAATTTCTAAACC...   \n",
       "scaffold_300679  CTCATTGCGTAAAGATAATAATATCCGTGTTTAAAAAAAAGATTGT...   \n",
       "scaffold_290644  TGTTTTAATTTTTCGGATGTAATAATAGCATCAGTATCACAATAAG...   \n",
       "scaffold_267693  TGTTTGATATGCTTTAATAACAATATTCTCTATTTGCTCTCTGTCA...   \n",
       "scaffold_241552  ATATATGTTCATATTTGGAATTTTTGGATGTATTTTAGGATATATT...   \n",
       "scaffold_231049  GAAAGAGAAAGTAGAAAAGAAATTTGATGGAAATGTAGTTGTATGG...   \n",
       "scaffold_58476   CTCCAATTTTAAAAAATCAATATCAGCTATTTTAATTTCTATTCCT...   \n",
       "scaffold_51309   AACTGTAAATGTTGTAGTGGCTGTTCCATCTCGTCTGTAATACTCA...   \n",
       "\n",
       "                                                     reverse_seq_r  \n",
       "scaffold_346753  CCTGTTATTGTTTCACAATTAATTGAAGGAATAAATTTATTTAAAT...  \n",
       "scaffold_338320  CTGAAAGGAGAGGATTAAATTTTGAAATTTTTGCGGTTATAACACC...  \n",
       "scaffold_300679  AAATTAAAAAAGCTAAAGCTGTCATTGATAAGGTTTTACCTTTACC...  \n",
       "scaffold_290644  TGTCTTAAACACGTAAATATTCTTTTATCAGGATTATCAAAATCAT...  \n",
       "scaffold_267693  AATTTTTTAATATATTTCATTTAATCATCCTTTTTATTTTTTTAAT...  \n",
       "scaffold_241552  AAAAAAAGATAATAAATATTTAGTTTTGAATATTGATGATGAAATT...  \n",
       "scaffold_231049  GATGGAGTGAAGAAGTATTTTGTGTTAGATGGTTGTTGGAGATGTA...  \n",
       "scaffold_58476   TCTTGTTTTTTATTAACAGAAGATAAATTATCAATAGCTACAACAT...  \n",
       "scaffold_51309   GTAAGAGTAACGAAATTATTATACCAAAAACATCTATGTCTTAATT...  \n",
       "\n",
       "[9 rows x 24 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df.sort_values('size')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2d73b5b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3891"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efeeb9d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
